### ISSUE TITLE
[discussion] Structure of bookmarks/highlights/clippings data

### ISSUE THREAD

**Issue by mustafa-001** [koreader/koreader#5794]
While previously working on Evernote exporter I've come to see there is couple of unimplemented to dos and broken functionalities. Here is couple of them and my opinions:

1- Exporting functionality depends on two different source; highlights and bookmarks. In the past I've suggested moving everything to bookmarks and using highlights as fallback for no-text-only-image clippings. But `bookmarks` lacks page number info and changing `bookmarks` creation code is little tricky. See below point 4 for more info on this. [little discussion about this](https://github.com/koreader/koreader/pull/5774#issuecomment-575890933)

2- `clippings` structured with chapter name but for now all notes indexed with `chapter` equals `1`. While adding this info to `bookmarks` is easy, additional nesting makes sorting tricky. For now it's like this:
`table.sort(book, function(v1, v2) return v1[1].page < v2[1].page end)`
but if chapter  where not to be `1` it would require sorting function to flatten data before to be:
` function(v1, v2) 
	local p1, p2
	for _, c1 in pairs(p1) do
		p1 = c1.page
	...
	return p1<p2` in top of having unnecessary additional for loops in exporter functions. Advantage of not "fixing" this is not having to worry about existing evernote.metadata file. Having nested structures in Lua tables if we don't need to select with that info (for example the need to select bookmarks of page vs to select clippings of chapter) is unnecessary complication to me. 

3- Title and author of exported book notes are parsed from file name, and if file isn't formatted with calibre or similar tool output format gets messy. Fixing this is easy, just use `doc_props` from book metadata. Caveat here is existing clippings indexed with old info, new exports won't overwrite old ones.

4- `bookmarks`'s  `page`  is pre-calculated value for `page = if self.has_pages and xpos0.page or xpos0` . 
So it's actually duplicate info and gives two different types to a variable. Puts responsibility of deciding how to interpret it to called function, while similar behaviour can be achieved with only passing `pos0` with same amount of if checks. [Bookmark code has check for it 12 times](https://github.com/koreader/koreader/blob/2736661bfcf6b472da8fa06202cb0f5f92013180/frontend/apps/reader/modules/readerbookmark.lua#L319)
I'm bringing this not to change existing code(not worth it while it's working), but to ask if I were to add "actual page info" as seen by user how should I rename new value and should I add it in the first place? 

5- Creation of a `bookmark` done from `ReaderHighlight`, it's confusing at best, not much problem. The thing is this `bookmark` not duplicated from `highlight`but recalculated from selected position info. And this can cause different `datetime` value and cause mismatches when finding pair of a highlight in `bookmarks`. 
[in code](https://github.com/koreader/koreader/blob/2736661bfcf6b472da8fa06202cb0f5f92013180/frontend/apps/reader/modules/readerhighlight.lua#L1088)

6- Bookmarks field has `notes` and `text` fields, last being modifiable one, this leads to storing same thing twice. If it were to me, I would implement one for static text(`notes`) and one for users comments about it(`comments`). This I think KoHighlights approach to this too. And this would make it easy and intuitive for adding notes unrelated to specific part of book. 

7- Storing user readable `datetime` instead of timestamp  makes sorting by creation complicated. Instead storing timestamp at file and converting it to user readable format would be easier to with it.

8- `clip.lua` has need for refactoring, there is only 4 functions off actual importance, others are confusing Kindle specific code.

9- `highlights` table holds no more information than `bookmarks` other than `pboxes`. Maybe unify them?

1, 2, 3 has more importance than others here and fixes for them easy but migration from existing table structures is what makes them hard. But while dealing with them maybe we can also improve existing code little better instead of carelessly patching things till they work. If we need a migration of any scale maybe we can do them all at once. Or all this just doesn't worth effort and I should give up and make my own text export plugin :).

So what are your opinions?
related https://github.com/koreader/koreader/issues/3415#issue-267829294
https://github.com/koreader/koreader/issues/1095#issuecomment-458918504
maybe this too https://github.com/koreader/koreader/issues/4951

**Comment by Frenzie** [comment]
I've never looked at this too much myself, so thanks for the post. Could there be performance-related concerns about merging it all in one big file if there were a lot of notes or highlights? (Probably less likely for bookmarks.)

Pinging @noembryo @embryo10 who may well know more about at least some aspects of this stuff than most of the active maintainers.

> Bookmark code has check for it 12 times

Yeah, that's probably just grown organically without being refactored or something.

**Comment by poire-z** [comment]
> maybe we can also improve existing code little better instead of carelessly patching things till they work. If we need a migration of any scale maybe we can do them all at once. Or all this just doesn't worth effort and I should give up and make my own text export plugin :).

I'm usually conservative and try to avoid changes for the sake of changes or when they are not really needed, but in this case, I wouldn't mind having bookmarks & highlights structure/code/storage reworked :)

Also, I don't know much about PDF/DJVU highlights, that pboxes stuff - and didn't even know we could highlight images (but I just tested that, and did not manage to! so, may be it's some obsolete or never implemented stuff?).
Also, I don't know  about evernote and the evernote plugin, how it works, what it expects, what's the output - and how it handles knowing past done stuff vs newly added stuff. I've never seen your `clippings` data structure :) (I wondered when reading your post, if by clipping, you always meant evernote clipping in the evernote context - or if some of them meant `highlights` in the plain reader context.)
But I know quite a bit more about CRE/EPUB/HTML bookmark and highlights.

So, some various remarks:

If we're ok with having a migration, we might as well think about a single complete and versatile data structure.
We have existing `highlight` (wrongly singular) and `bookmarks` (plural). Dunno how we could name the new single one... `highlights` (plural), `favorites`,... ? There must be a generic word for that we'd just have to find :)

> If we need a migration of any scale maybe we can do them all at once

That would be complicated to do, as we don't know where all the user books might be (spanned in multiple directories, some may have been backup'ed and will be copied with their .sdr on the device later). Lot easier and leaner to just convert them when loading a book for the first time with a KOReader version that will include this change and the migration code. We have [settings_migration.lua](https://github.com/koreader/koreader/blob/master/frontend/ui/data/settings_migration.lua) that could deal with that: when the new setting is not there, create it and migrate/merge data from the old `bookmarks` and `highlight`.
Then, either keep the old data structure (in case the user downgrade to a previous version, he could see its past highlights), or just backup the previous [metadata.lua](url) to `metadata.lua.old_bookmarks` when making the migration (and forget about them).
@noembryo would need to update KOHighlights to use the new one if present, or keep using the old ones when not.

So, if you're having a go at that, I would just
1) think about the correct and future-proof data structure we want, with slots for all the things we need. May be with or without the per-page-number first level that `highlight` has. May be as an array (adressed by useless numbers, the ones the first `for` loop skips), or as a table (adressed by some unique id), with possibly some other keys that could store some kind of index/orderered list of the original IDs for quicker sorting/adressing, if we want it saved (if not, we could just build that in memory at book opening time).
About storing the human readable date: a timestamp would be fine, but I guess we could store some ISO human readable one, and alphanumericaly-sortable and easily parsable. But what we should no more do is storing the date and page number in the text! We need to have individual parts stored as different keys: the selected text as `text` or `selected_text`, the user comments or editiedtext as `notes` or `comments`, the date of selection as `selected_date`, the date of modification or addition of notes as `notes_date`, the page number (first page? multiple pages if selection spans multiple page?). Should help with [some issue in RTL ](https://github.com/koreader/koreader/issues/5359#issuecomment-570484948), but for anything translated, we should not hardcode it in the values: the word "Page" and the formatted date should not be in `text` - if the user switches languages, we don't want to find there words from the old language.
And some flags to distinguish between page bookmarks (no text, but allow notes?), highlights and their drawing method,...
This is the most interesting part :)
2) rewrite ReaderHighlight, ReaderBookmark, ReaderView, and the Evernote plugin to fit that new datastructure, and just forget about the old code
This might be fun if you made the right choices for the datastructure :)
3) do the migration stuff
This is less fun :)

> 9- highlights table holds no more information than bookmarks other than pboxes.

It also contains `["drawer"] = "lighten"`, the way to highlight it: gray box, black box, or underlined.
This might also be changed: dont store the drawing method, but the user-set type for the selection: comment, correction, and associated a drawing method to a selected text type. Dunno.

> 4- bookmarks's page is pre-calculated value for page = if self.has_pages and xpos0.page or xpos0 .
So it's actually duplicate info and gives two different types to a variable. Puts responsibility of deciding how to interpret it to called function, while similar behaviour can be achieved with only passing pos0 with same amount of if checks. Bookmark code has check for it 12 times
I'm bringing this not to change existing code(not worth it while it's working), but to ask if I were to add "actual page info" as seen by user how should I rename new value and should I add it in the first place?

I think it deserves its own slot. It's fixed for PDF, but it may be updated for CRE when the user changes font size (we actually fetch it on each page turn to know if a bookmark is on the current page, so it must be cheap to keep it updated.
Note that page numbers with EPUB/HTML documents may change when the user changes options like font size: we currently show in the text the old page number it was made on, - which has no sense and is wrong if the user changed font size...
That's why we're only using xpointers for CRE, and why there are all these checks:
```lua
if self.ui.document.info.has_pages then
    -- PDF, pages, pboxes
else
    -- CRE, xpointers, variable pages, segments
end
```
the handling are quite different, so you might have a hard time avoiding them.

About the performance: I think it should be fast in the Reader* code, as it may be sollicitated on each page turn (or at book loading time, where it would be allowed to be slower). I'd be fine if it is slower in the exporting plugins, which one uses only occasionally.

> Could there be performance-related concerns about merging it all in one big file if there were a lot of notes or highlights? (Probably less likely for bookmarks.)

They are all already in one single (big if many highlights and bookmarks) file: our metadata.ext.lua. So, merging them will make that single file smaller :)

I guess we could think of using LuaData.lua for the new datastructure, but that might add more complexity to something that's going to be already complex (think CRE updated and cached page numbers updated on each font size changes > everything re-appended to the LuaData). But it's an option (not too fan of it, I always have a hard time getting into how it works, and debugging stuff when it is involved).

OK, I'm sure I'd have dozens more questions and suggestions if I were to actually think deeper about that - so, holding on :)

**Comment by Frenzie** [comment]
> (but I just tested that, and did not manage to! so, may be it's some obsolete or never implemented stuff?).

I imagine MuPDF can; I don't think KOReader can but I'm not a heavy highlighter.

**Comment by noembryo** [comment]
While writing a comment, @poire-z posted his very comprehensive (as usual ;o) comment and made mine obsolete..
Some things I might add:
- I do agree that a rethinking of the storage for highlights/bookmarks would be very nice.
Alas it _will_ be some more work for the KoHighlights to keep the compatibility (it already tries to read metadata from the 2014 version of KOReader)
- If a merging is going to happen, I would suggest to create a totally new key (`highlights` seems fine), so that we can check for its existence if we are to parse the file in a new way.
This also let us keep the older keys of an already opened book for backwards compatibility.
- The page numbering is problematic because it is using the current page.
This can be anything, after a font or device (in case of syncing) change.
Maybe a percentage could help more? Or an absolute page number based on book's word count and percent (just saying ;o)
- No problem with the current date format (for sorting or anything)
I could convert it if needed, but it doesn't.

**Comment by mustafa-001** [comment]
> and didn't even know we could highlight images

When parsing images gets serialized to .png and stored in `clipping`. But text, Joplin exporters ignore it, not sure about code on html and Evernote but last I checked it didn't work with html too.

>  I've never seen your clippings data structure :)

It's an intermediate format, centralized for all books, parsed from `highlights´, used for exporter functions. External programs also can use it.

>That would be complicated to do, as we don't know where all the user books might be

Yes, you're right, I meant doing all changes in one release.

>About storing the human readable date: a timestamp would be fine, but I guess we could store some ISO human readable one, and alphanumericaly-sortable and easily parsable.

I think only a timestamp would be fine, if it needs to be in human readable format for display or export it's not a heavy calculation. If we only have a `creation_time` having both format is only a storage size problem but with `last_modification_time` things would get messy easily. 
> We need to have individual parts stored as different keys

Exactly, we shouldn't edit copied text from book. People don't change sentences in books, they write notes on corners of a book. There were suggestions/requests about how bookmarking should work in older issues, we could look at them also.



Something I would definitely want is having less deeper tables, we shouldn't index by chapter or page. For example:

```
`annotations`
	|--`note_number`
	  |-`text`
	  |-`comment`
	  |-`chapter`
	  |-`xpos0`
```

not

```
`annotations'
	|-`chapter`
		|-`page`
		  |-`note_number`  
			|-`text`
			|-`comment`
```


>Note that page numbers with EPUB/HTML documents may change when the user changes options like font size: we currently show in the text the old page number it was made on, - which has no sense and is wrong if the user changed font size...

I get this, but if we want to compute position for sorting purposes we would just use `xpos0`, 

```
if self.ui.document.info.has_pages then --this selector can be type(xpos9) == string
    -- PDF, xpos0.page
else
    -- CRE, getPageFromXPointers(xpos0)
end
```

or if we want to store something sortable without any calculation we can pre-calculated page number for fixed font size for only sorting purposes or percentage as @noembryo said. 

```
if self.ui.document.info.has_pages then
    -- PDF, sorting_number = xpos0.page
else
    -- CRE, sorting_number = getPageFromXPointersWithFontSize(xpos0, 25)
end

```
Currently we aren't doing anything extra with it, it has potential to be more useful and clear. I wouldn't expect a variable named `page` to contain xpointer.



>I'd be fine if it is slower in the exporting plugins, which one uses only occasionally.

It would be faster for exporting too since current code indexes both tables. I don't have any reason to expect performance drop from unified table.

**Comment by poire-z** [comment]
> When parsing images gets serialized to .png and stored in clippings

The [only trace](https://github.com/koreader/koreader/commit/566586619#diff-483d4f2c6d1ec179b6578caf247d520d) of the word "clipping" in the frontend/ code has it being commented in 2014.
But ok, I see that's what the evernote plugins does itself when the highlight has an empty text="":
https://github.com/koreader/koreader/blob/dfd2baa3adcd1f6126e7a4e6b2e928499b9b08ca/plugins/evernote.koplugin/clip.lua#L244-L256
`MyClipping:getImage()` opens the document, and do something like a screenshot of the area bounded by these coordinates :) which might only work on PDFs, not on EPUBs.
So, I'd consider that a bit external to the bookmarks/highlights rework :) unless you want to make it more part of it.

> I meant doing all changes in one release.

Agreed, we'd have to get it right on its first merge :)

> I think only a timestamp would be fine, if it needs to be in human readable format for display or export it's not a heavy calculation. If we only have a creation_time having both format is only a storage size problem but with last_modification_time things would get messy easily.

Well, when debugging, you (and I) would be happy to see a `2020-01-28-14:32:23` to locate if what's just been added after our fix is ok in the file :) But ok on the principle of storing a ts `1580226170`, may be with another (never used and not advised to be used) string representation of it in another key.

> we shouldn't index by chapter or page

Agreed with the flat layout. A one-level for loop can easily filter those that have the `page=` that it's interested in.

> we would just use xpos0

Fine with me.

> or if we want to store something sortable without any calculation we can pre-calculated page number for fixed font size for only sorting purposes or percentage
getPageFromXPointersWithFontSize(xpos0, 25)

Nope, that won't do, it would need a re-render of the full book with that font size, which might takes minutes - and re-render it again to get back to the user selected font size.
I don't like percentage either, and you'd need some fractional part for the user to be able to guess which highlight comes after another, not really user friendly.
I'd just update the stored `page` everytime we call getPageFromXPointer (which can and will be cached with #5773, so: cheap) - so the user would see the page number on the latest device and settings he has been using.

So, I guess we agree on the target: less deep tables, more keys to store more individual thingies.

`annotations` might be a candidate for a name for that new datastructure: it is different enough and would avoid confusion when discussing it :) if we accept that a bookmarked page is some kind of annotation - and that non-annotated text selection is also an annotation :|
Or just `marks`? (but it won't help with `grep marks` :|)

**Comment by poire-z** [comment]
For the individual annotations/marks, we'll need to think about what's different between EPUBs and PDFs, what is common, what's required for each.
Quick idea:
common stuff flat at first level, and may be the technical position/location/way to reach it in the book in a subtable (that KOHighlights/exporter could just ignore):
```lua
annotation[123] = {
  text = "What the book author wrote",
  notes = "My comments on it",
  chapter_name = "1.1 Nature de l'empire",
  creation_ts = 1580226170,
  creation_date_do_not_use = "2020-01-28-14:32:23",
  page = 45, -- whether fixed in PDF, or updated from the latest font size used
  percent = 13.23543432423 -- if it can help for quick (re)ordering, although that may change with EPUB rendering layout/CSS styles change
  type = "selection/correction/annotation",
  is_page_bookmark = true/false, -- of is_highlight = true/false
   -- currently, we have ["highlighted"] = true,
  is_page_clip = true/false, -- when no text, so a screenshot required
  drawing = "underline", -- nil or false is is_page_bookmark?
  location = {
    -- for CREDocuments:
    pos0 = "/body/DocFragment/body/div/p[28]/text()[1].0",
    pos1 = "/body/DocFragment/body/div/p[28]/text()[3].73", -- absent for page bookmarks?
    -- for PDFDocument and others
    all the pboxes/boxes stuff, or empty for page bookmarks?
  }
```

Bookmarked page durrently do have `["text"] = "Page 15 in 1 Nature de l'empire @ 2019-12-07 21:25:49"`. Would we store the chapter name in `text=`, or just add another key for all:
`chapter_name = "Nature de l'empire",`

**Comment by noembryo** [comment]
Annotations is fine.
`Annotate: add notes to (a text or diagram) giving explanation or comment.`

So we can:
- Annotate a book
- Annotate a page (bookmark)
- Annotate a sentence/word (highlight)

**Comment by noembryo** [comment]
I will try to monitor the progress on this, but in case its happening somewhere else and I missed it:
I would really appreciate it, if I could get some samples of the new format (when finalized), so that I could adapt it for KOHighlights.
It would be nice to have an update out that supports it, before its release :o)

**Comment by mustafa-001** [comment]
So far I've managed to get `Readerview` and `ReaderHighlight` refactored (I've been little busy :') ). 

This is the structure I used:
```lua
`["annotations"] = {
        [2] = {
            ["page"] = 10, --according to fontsize when this annotation created in CRE documents --:getPageFromXPointer()
            ["chapter"] = "SCENE I. Verona. A public place.",
            ["text"] = "quench the  on pain of death, all men depart.Exeunt all but MONTAGUE, LADY MONTAGUE, and BENVOLIOMONTAGUE",
            ["drawer"] = "lighten",
            ["creation_timestamp"] = 1580923107,
            ["pos0"] = "/body/DocFragment[5]/body/div/div/div/blockquote[52]/p[1]/text()[4].6",
            ["total_page"] = 134, --according to fontsize when this annotation created for CRE documents --:getTotalPages()
            ["percent"] = 0.074626865671642,
            ["pos1"] = "/body/DocFragment[5]/body/div/div/div/p[52]/strong/text().8",
            ["datetime"] = "2020-02-05 20:18:27"
			["comment"] = "TODO",
        },`

```

Also I'll add `is_highlight` (for deciding should `ReaderView` attempt to draw it, maybe a more obvious name?) and `is_picture`. 

I didn't used percent, this way:

- We can sort by `page/total_page`, also no need to change logic for fixed page documents.
- Exporters can get both fixed page info and page info relational to other annotations. If need arises one can simply make all `total_page` values equal for a book. Similar to storing 	`percent` and `page` or `percent` and `total_page` but little less backwards.

I still need to refactor dictionary lookup and bookmarks and have couple of questions. Tests and immigration is another pain :).

  > `percent = 13.23543432423 -- if it can help for quick (re)ordering, although that may change with EPUB rendering layout/CSS styles change`

Since we get page info from XPointer is this possible?

>  `type = "selection/correction/annotation",`

Is correction necessary? My idea was that you just add a comment to existing annotation and that comment is the text that shows up in `Rename` option on highlight menu.  You can add a empty text annotation by adding bookmark and editing it's comment field from Bookmarks menu. Not sure about what annotation and selection exactly is.

How to name functions in `ReaderHighlight` and `ReaderBookmark`? My idea is to left them unchanged at least for now. For example `onShowHighlightMenu` is still used for highlighting related info. 

This is what I managed to do so far, https://github.com/mustafa-001/koreader/commit/58882a43510a3278708be560eeae5dbc8c60a833. It's for now little messy but when all idea gets concrete I'll make a new branch with more explanatory commit history.

**Comment by poire-z** [comment]
> I've been little busy

Take your time, no hurry :)

Looks fine so far. You're mostly getting rid of the [page] level. Everything working fine (as before) with just this?
Added a few comments on your commit (don't rebase it before reading them, I think I once lost some comments @Frenzie did, and could only find them in my mail :)
Mostly, I'd prefer if things were kept named xxxHighlightyyy when we're _drawing highlights_, even if we're using the `annotation` datastructure.

(Some of the strange things you may have wondered about came from https://github.com/koreader/koreader/pull/3369.)

So, you'll be using `creation_timestamp` as the unique identifier for an annotation. Which looks probably fine (I'd be more confortable with some fractional digits :) but any stuff migrating or adding annotations in batch should take care of faking them and avoid identical ones.
(Dunno if we need some kind of other unique ID - but a timestamp is indeed more readable and concise :)

> Also I'll add is_highlight (for deciding should ReaderView attempt to draw it, maybe a more obvious name?) and is_picture.

Might as well have one for each of them (to avoid checks for `if not is_highlight and not is_picture`): `is_bookmark, is_highlight, is_picture`. `is_highlight` is fine, we'll still be drawing highlights, stored in annotations :)


> percent = 13.23543432423 -- if it can help for quick (re)ordering, although that may change with EPUB rendering layout/CSS styles change
Since we get page info from XPointer is this possible?

That would be (as used in ReaderRolling:getLastPercent(), but it would work also in page mode):
`self.ui.document:getPosFromXPointer(self.ui.document:getXPointer()) / self.ui.document.info.doc_height`

> type = "selection/correction/annotation",
Is correction necessary? My idea was that you just add a comment to existing annotation and that comment is the text that shows up in Rename option on highlight menu. 

Not necessary. That's just an available slot to give a "tag" to something selected. Would have no use for now, but could be used in the future if we want to allow a user a checkbox list after selecting so it could tag the selection he just made as being a "correction", "need to lookup that on the net", a "needs fact checking" ... :) just thinking about the future need for slots, no need to implement that yet :)
And ok, once text is edited, and so has a comment, it does not need to change the type. (although it could change from `selection` to `comment` once it has a comment and is not just a selection anymore :)

> How to name functions in ReaderHighlight and ReaderBookmark? My idea is to left them unchanged at least for now. For example onShowHighlightMenu is still used for highlighting related info.

Agreed. Let them deal with the part they manage, even if all are stored in the `annotations `datatstructure.

> We can sort by page/total_page, also no need to change logic for fixed page documents.
Exporters can get both fixed page info and page info relational to other annotations. If need arises one can simply make all total_page values equal for a book. Similar to storing percent and page or percent and total_page but little less backwards.

Go with that.
Later, we could maybe store `page .. y_in_page` for PDF document, and `pos_y / total_height` for CRE document, to be able to order highlights in a same page (where page/total_page would give the same number).
Also, and you can let that to me for later, for CRE, we could easily have the page and total_page checked/updated (on load and on re-rendering), so it's always accurate.

(We'll also want to have a more accurate `chapter` for a highlight than the one for the page - there may be multiple chapters on a page, best to use the one just before the highlight :) you can leave that to me too.)

**Comment by poire-z** [comment]
Cross ref (because GitHub stupid "Discussions" don't auto-do that...) : https://github.com/koreader/koreader/discussions/7268

> integrate the translations and the notes somehow, so I have these two ideas.
    Add an option in the translate box, either for:
    Copying the translation result text (to add it later as a highlight note), or
    For adding the translation result into a note text for the same highlighted text
    Also to see the note as soon as you tap in the highlighted text, in a non intrusive way. (or to have an option to activate this type of visualization on the settings)

**Comment by poire-z** [comment]
With the discussion and ideas in #8301, I'm tempted to take on this, and sign for a few weeks of vacation in hell :/
I haven't yet looked at the current code, so just throwing thoughts (from memory, which may be wrong):
- ReaderHighlight is more about text selection and what to do with what's selected, than really about highlights.
- ReaderView is about drawing in general, and drawing highlights in particular
- ReaderBookmark is mostly about bookmarks&highlights UI (list, edit, remove)
- there are exchange between ReaderHighlight and ReaderBookmark to keep duplications in sync that makes a good part of the ugliness of it all

We could have:
- a standalone AnnotationList module that would wrap what we save
- (it could be a ReaderAnnotation module, but may be it should be foreign to being a Reader module, so it can be used standalone by other stuff (like exporting, plugins...))
- this AnnotationList object could have Annotation objects, with may be subclasses: CreAnnotation and NonCreAnnotation, that would hide the Cre/Pdf pecularities, and avoid us to have too many of these `if document.info.has_pages` we get everywhere
- AnnotationList would have methods/services that returns a list of these Annotation objects depending on who (ReaderBookmark, ReaderHighlight, ReaderView) needs what (page bookmarks only, the highlights in the view for drawing...)
- Annotation objects would have methods to get pages, text, notes, formatted text for use in various contexts...

Might be more fun to start with having a nice and clean architecture, trash all the current bookmark and highlights code, and rebuild it bits by bits from there - and postpone caring about the migration until later when we're satisfied with some proper clean stuff.

**Comment by hius07** [comment]
Internally: is it necessary to store annotations sorted?
For annotations list it is not necessary, item_table is built in accordance with the required type of sorting.
Is it necessary for quick drawing a page with a dogear and highlights?
For dogear, the bottleneck is
https://github.com/koreader/koreader/blob/0d231cbbef487c2a83b4ebb939490ecbbb929163/frontend/apps/reader/modules/readerbookmark.lua#L455-L469
Can such search (with other algorythm) be rather fast for non-sorted annotations table?

**Comment by hius07** [comment]
> Are you really taking on this work ?

No

**Comment by hius07** [comment]
More categories

![1](https://user-images.githubusercontent.com/62179190/141938461-560c3446-7e6d-4e44-9fea-77c8a9a7c451.jpg)

**Comment by ichnilatis-gr** [comment]
I have given some ideas with screenshots from highlights management on the Pocketbook device [here](https://github.com/koreader/koreader/issues/8535#issuecomment-999681237). What do you think about the two separate options for editing highlight text and notes? Is something like this possible for KOReader?

**Comment by hius07** [comment]
> editing highlight text

What is a use case for this?

**Comment by ichnilatis-gr** [comment]
We can make changes/additions etc. to the highlighted text without it appearing as a note and without it being displayed twice (without and with changes, i.e. as original text and as note) in the exported file. It will be displayed only with the changes.

**Comment by hius07** [comment]
> If the user could tag its "note" as being a "correction",

I experimented with new highlight drawer "strikethrough", looks okay, can be used as indicator of correction (the corrected text visible in the popup note).
Didn't think about export yet.

**Comment by poire-z** [comment]
Regarding the ambitious goal outlined above, I must say that I've become quite frightened by the amount of work it would need :) There's been so many tweaks and fixes added these last months by @hius07 that it would make a lot to change and test.
Moreover, what we end up having looks quite good and working for users, and may be it's just good enough. And the fact @hius07 was able to build all these nice UI over these ugly datastructres means we don't really need to have nice datastructures :)

So, maybe we could lower our ambitions (I could :), and just clean up some of the mess ?
I'm thinking mostly about killing the duplicated `highlight`, and moving its properties (drawer, pboxes) into the corresponding `bookmarks` item.
And if we have to do a migration for that, we can just try fixing/removing now the hardcoded autotext prefixes and suffixes added, and make the datastructure as with `bookmarks_items_auto_text=false` and use it just for display and export.

I could be fine with the possibility of having only a single "note", and being able to mark what that note is (correction, translation).
We could add other properties to each `bookmark` item if needed.

The last real itchy thing for me is that bookmark.note is the original text, and bookmark.text is the user note... This hurts my stomach everytime I review @hius07 :) but fixing that would mean having to go over every and each single bit @hius07 updated these last months :/

@hius07: mentionning all this just to let you know that if you share some of these lower ambitions and you have been restrained by my above "stop! we-can-do-better", feel free to have a go at it :)

**Comment by ichnilatis-gr** [comment]
Let me supplement my [above thought](https://github.com/koreader/koreader/issues/5794#issuecomment-1001548871) about two editing options (for highlights and notes) with an example.
It happens that we make a highlight in a pdf file and it includes together with the last word a reference number, which we want to delete from the highlight text and in addition we want to add a note, so that when we export the highlights, both are exported (highlight text and note). So we need an option to edit the highlight and an option to add a note.

I'm sorry to make things complicated, it just relates to some difficulties we encounter in practice.

**Comment by hius07** [comment]
Technically I see no problems with editing bookmark text, it doesn't break any checks.
(Side effect: if the highlighted text within bookmark is changed, some bookmarks type might be changed from "highlight" to "note").
Should we still name it "highlight" after editing?

**Comment by poire-z** [comment]
> Technically I see no problems with editing bookmark text

You mean modifying the original source text (in the "notes" slot) ? If, yes, do we really want that ?
No idea what our ambitions and motivations could be in the future, but I think having the garantee that this slot is and stays the original book text is nice to have and keep.
(And allowing to edit the note + the bookmark original text would not be obvious in the UI.)

(We don't have to answer all requests if they don't fit in the current model. For me, this one did not fit - and it could only be postponed until we have something better to work on. I'm afraid of pushing the franskesteinasation of our bookmarks too far :)
But trusting you on all things bookmarks.

But @ichnilatis-gr request is more/just about the result after exporting, so it could as well be just a flag in the bookmark indicating we don't want to export the original text.

**Comment by ichnilatis-gr** [comment]
> Should we still name it "highlight" after editing?

Why not? It still remains a highlight with some changes/additions. I've given some ideas [here](https://github.com/koreader/koreader/issues/8535#issuecomment-999681237).


> But @ichnilatis-gr request is more/just about the result after exporting, so it could as well be just a flag in the bookmark indicating we don't want to export the original text.

It's not just about the result after exporting; there may be an addition/correction to the original highlighted text (which in a pdf may not be correctly recognised) to which we may want to add a note.

**Comment by ichnilatis-gr** [comment]
Sorry to come back; I was just wondering if there is any hope of creating a button for [that option I suggested](https://github.com/koreader/koreader/issues/5794#issuecomment-1008032154)?

**Comment by uroybd** [comment]
> Regarding the ambitious goal outlined above, I must say that I've become quite frightened by the amount of work it would need :) There's been so many tweaks and fixes added these last months by @hius07 that it would make a lot to change and test. Moreover, what we end up having looks quite good and working for users, and may be it's just good enough. And the fact @hius07 was able to build all these nice UI over these ugly datastructres means we don't really need to have nice datastructures :)
> 
> So, maybe we could lower our ambitions (I could :), and just clean up some of the mess ? I'm thinking mostly about killing the duplicated `highlight`, and moving its properties (drawer, pboxes) into the corresponding `bookmarks` item. And if we have to do a migration for that, we can just try fixing/removing now the hardcoded autotext prefixes and suffixes added, and make the datastructure as with `bookmarks_items_auto_text=false` and use it just for display and export.
> 
> I could be fine with the possibility of having only a single "note", and being able to mark what that note is (correction, translation). We could add other properties to each `bookmark` item if needed.
> 
> The last real itchy thing for me is that bookmark.note is the original text, and bookmark.text is the user note... This hurts my stomach everytime I review @hius07 :) but fixing that would mean having to go over every and each single bit @hius07 updated these last months :/
> 
> @hius07: mentionning all this just to let you know that if you share some of these lower ambitions and you have been restrained by my above "stop! we-can-do-better", feel free to have a go at it :)

I would very much like to clean up this mess if you allow me to do so. Exporting will be so much easy If we do that.

**Comment by poire-z** [comment]
> I would very much like to clean up this mess if you allow me to do so

Unless @hius07 has already started some work about it, or have other thoughts about the whole issue, go for it.
You got familiar with highlights and already tackled a one-time-migration in #8996 (which I guess should be closed unmerged as it was not a good idea ?), you probably know more about the "exporting" side of bookmarks/highlights than any of us - and you seem to be courageous and to take our severe reviews quite well :) so, you might be _The One_ for the task :)
(It will also need some courage and time from some of us to review, and the scrutiny of @hius07 :)

> Exporting will be so much easy If we do that.

Probably keep that for after #8944 is merged (even if not fully nice) - to keep having smaller independant PRs :)
May be take some time in the meantime to read this whole discussion - may be you'll have some new/other ideas about how to get to something even better, in small independant steps.

**Comment by uroybd** [comment]
> which I guess should be closed unmerged as it was not a good idea?

Yes, it should go unmerged.

By brainstorming with @hius07, I believe we can plan a good, extensible, forward-looking merging of highlights and bookmarks that will allow dependent parts like exporter to be more accurate, and concise without much manual intervention like horrible lopping and redundant nested lists to clean up.

**Comment by poire-z** [comment]
I don't think there was a link in the first post, but for reference, the original work by @mustafa-001 : 
https://github.com/mustafa-001/koreader/commit/58882a43510a3278708be560eeae5dbc8c60a833
It's a 2 years old commit, so it's probably not reusable, but it is killing the nested loop.

**Comment by ichnilatis-gr** [comment]
Dear @WolfgangDpunkt, @hius07 (glad to know you are well), @offset-torque, @zwim, @mergen3107, @NiLuJe, @Frenzie and @poire-z, thank you all for your contribution in issues #9356 and #9395. It's a feature I've been wanting too, but I didn't dare ask for it.
If I am not being annoying (I know I am :-)) I would like to remind you about my [request above](https://github.com/koreader/koreader/issues/5794#issuecomment-1008032154).

Is such a thing possible?
Thanks

**Comment by ryanwwest** [comment]
Mentioned this in https://github.com/koreader/koreader/pull/9031 to ask about the Highlight refactor PR progress. not sure why it isn't showing up on its own...
